### PR TITLE
Ensure volumes are unmounted during graceful node shutdown

### LINKED ISSUE #115148: Graceful node shutdown doesn't wait for volume teardown
### What happened?

We have observed that after a node is preempted, pods using persistent volumes take 6+ minutes to restart. This happens when a volume was not able to be cleanly unmounted (indicated by updating `Node.Status.VolumesInUse`), which causes the Attach/Detach controller (in kube-controller-manager) to wait for 6 minutes before issuing a force detach. In addition to a long pod restart period, an unclean mount could also cause data corruption.

From the logs we've observed the following non-ideal behaviors, which I think reduces to the same issue that we only check for containers' being terminated and not their volumes.

1. The CSI driver running on the node gets terminated before the workload's volumes get unmounted. This causes unmount to fail. In our case, the CSI driver is part of the `system-node-critical" class, but since we don't wait for volumes to be unmounted, we move on to shutting down the critical pods.

3. The OS gets shutdown before volume teardown is completed, even if there is still time left in the `shutdownGracePeriod`. In our case, the `shutdownGracePeriod` is set to 30 seconds, however we see that the node got shutdown in 10 seconds, even while unmount has not succeeded.

kubelet.log
```
Jan 17 21:48:18 ip-10-0-1-208 kubelet[950]: I0117 21:48:18.460409     950 setters.go:548] "Node became not ready" node="ip-10-0-1-208" condition={Type:Ready Status:False LastHeartbeatTime:2023-01-17 21:48:18.46035065 +0000 UTC m=+1235.025577092 LastTransitionTime:2023-01-17 21:48:18.46035065 +0000 UTC m=+1235.025577092 Reason:KubeletNotReady Message:node is shutting down}
...
Jan 17 21:48:27 ip-10-0-1-208 kubelet[950]: E0117 21:48:27.126414     950 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/csi/ebs.csi.aws.com^vol-00477959568e1db98 podName: nodeName:}" failed. No retries permitted until 2023-01-17 21:48:35.126400526 +0000 UTC m=+1251.691626967 (durationBeforeRetry 8s). Error: UnmountDevice failed for volume "pvc-513e457c-86c1-4564-911c-6c0670cb839c" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-00477959568e1db98") on node "ip-10-0-1-208" : kubernetes.io/csi: attacher.UnmountDevice failed to create newCsiDriverClient: driver name ebs.csi.aws.com not found in the list of registered CSI drivers
...
Jan 17 21:48:29 ip-10-0-1-208 kubelet[950]: E0117 21:48:29.006761     950 kubelet.go:2475] "Container runtime network not ready" networkReady="NetworkReady=false reason:NetworkPluginNotReady message:Network plugin returns error: cni plugin not initialized"
Connection to 10.0.1.208 closed by remote host.
Connection to 10.0.1.208 closed.
```

Some code tracing:

1. The logic for releasing the inhibitor lock waits for pods to be terminated within the grace period. 

https://github.com/kubernetes/kubernetes/blob/37e73b419e455db34f5fe3e8d815418680ab23df/pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go#L326

2. It depends on `IsTerminated` to indicate that the pod has been killed: https://github.com/kubernetes/kubernetes/blob/37e73b419e455db34f5fe3e8d815418680ab23df/pkg/kubelet/pod_workers.go#L676

3. `IsTerminated` depends on `terminatedAt`, which is set when all containers have stopped:
https://github.com/kubernetes/kubernetes/blob/37e73b419e455db34f5fe3e8d815418680ab23df/pkg/kubelet/pod_workers.go#L273

4. However, volume termination for a Pod only begins when all containers have stopped:
https://github.com/kubernetes/kubernetes/blob/37e73b419e455db34f5fe3e8d815418680ab23df/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L249

### What did you expect to happen?

Node graceful shutdown to also wait for volumes to be torn down.

### How can we reproduce it (as minimally and precisely as possible)?

Deploy a StatefulSet pod using an attachable PVC, and then trigger node shutdown. 

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

Tested on 1.25, 1.26

</details>

### Cloud provider

<details>
Observed on GCP, AWS
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by torredil**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR ensures that CSI drivers have the opportunity to perform necessary volume teardown during node shutdown, preventing data corruption and reducing pod restart times.

The current node shutdown manager and volume manager in Kubelet both operate off the same signal (termination of pods), which leads to the following race condition:

image
_Diagram as provided by @msau42_ 

#### Overview of changes
`VolumeManager` has been added as a dependency to the node shutdown controller (`nodeshutdown.Manager`) in Kubelet. This change allows for querying the status of volume mounts before proceeding with the termination of pods. Waiting for unmounts is done on a best effort basis - if the volume unmounting process fails, the error is logged, but the shutdown process continues.

#### Which issue(s) this PR fixes:

Fixes #115148.

#### Special notes for your reviewer:

This PR builds on the work started in [#120091](https://github.com/kubernetes/kubernetes/pull/120091). Credit to [@mauriciopoppe](https://github.com/mauriciopoppe) for authoring the initial fix.

#### Does this PR introduce a user-facing change?

```release-note
Node shutdown controller now makes a best effort to wait for CSI Drivers to complete the volume teardown process according to the pod priority groups.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

#115148

```docs

```

### CODE REVIEW

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go [hunk 1]
```diff
@@ -394,22 +399,47 @@ func (m *managerImpl) processShutdownEvent() error {
...
 			}(pod, group)
 		}
 
+		var groupTerminationWaitDuration = time.Duration(group.ShutdownGracePeriodSeconds) * time.Second
 		var (
-			doneCh = make(chan struct{})
-			timer  = m.clock.NewTimer(time.Duration(group.ShutdownGracePeriodSeconds) * time.Second)
+			doneCh         = make(chan struct{})
+			timer          = m.clock.NewTimer(groupTerminationWaitDuration)
+			ctx, ctxCancel = context.WithDeadline(context.Background(), time.Now().Add(groupTerminationWaitDuration))
```
Comment (gnufied) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:410]: nit: I would use `context.WithTimeout`

**Comment by gnufied** [comment]
lgtm from storage side. We really need to get input from @mrunalp or someone from node team, who actually designed and implemented graceful node-shutdown feature.

**Comment by torredil** [comment]
The following PRs need to be merged to exercise the end to end test: 
- https://github.com/kubernetes/cloud-provider-gcp/pull/721
- https://github.com/kubernetes/test-infra/pull/32728

**Comment by torredil** [comment]
The test failure in `pull-kubernetes-e2e-gce` is expected and can be ignored (graceful node shutdown is not enabled for that job). I tried to disable the test from executing there by adding the `[Feature]` tag but ran into the following: 

```
ERROR: k8s.io/kubernetes/test/e2e/storage/graceful_node_shutdown.go:60: [Feature:GracefulNodeShutdown] in plain text is deprecated and must be added through WithFeature(GracefulNodeShutdown) instead
```
`WithFeature(GracefulNodeShutdown)` adds the following prefix: `NodeFeature` hence why the test is not skipped there. I have addressed that here: https://github.com/kubernetes/test-infra/pull/32728/files#diff-6c025406cbaaee64363952c64c26d7c09f183c0345c55fd8cd7a3384ede2aecfR346

File: pkg/kubelet/volumemanager/volume_manager.go [hunk 2]
```diff
@@ -106,6 +108,12 @@ type VolumeManager interface {
 	// the duration defined in podAttachAndMountTimeout.
 	WaitForUnmount(ctx context.Context, pod *v1.Pod) error
 
+	// WaitForAllPodsUnmount is a version of WaitForUnmount that blocks and
+	// waits until all the volumes belonging to all the pods are unmounted.
+	// An error is returned if there's at least one Pod with volumes not unmounted
+	// within the duration defined in podAttachAndMountTimeout
```
Comment (kwilczynski) [inline pkg/kubelet/volumemanager/volume_manager.go:114]:
```suggestion
	// within the duration defined in podAttachAndMountTimeout.
```

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go [hunk 3]
```diff
@@ -394,22 +399,47 @@ func (m *managerImpl) processShutdownEvent() error {
 				} else {
 					m.logger.V(1).Info("Shutdown manager finished killing pod", "pod", klog.KObj(pod))
 				}
+
```
Comment (kwilczynski) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:402]: Why the extra newline?

File: pkg/kubelet/volumemanager/volume_manager.go [hunk 4]
```diff
@@ -479,6 +487,23 @@ func (vm *volumeManager) WaitForUnmount(ctx context.Context, pod *v1.Pod) error
 	return nil
 }
 
+func (vm *volumeManager) WaitForAllPodsUnmount(ctx context.Context, pods []*v1.Pod) error {
+	var errors []error
+	var wg sync.WaitGroup
```
Comment (kwilczynski) [inline pkg/kubelet/volumemanager/volume_manager.go:492]:
```suggestion
	var (
		errors []error
		wg sync.WaitGroup
	)
```

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go (hunk 3 above)
Comment (torredil) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go]: Fixed.

**Comment by kwilczynski** [comment]
Looks good!

However, it would be nice to get a review from @mrunalp, @dims, and @SergeyKanzhelev, if they have some time.

**Comment by rphillips** [comment]
Ideally, I believe we will want to wait for the readiness of the pod to go to false before trying to teardown the volumes. This will allow the pod to gracefully terminate without racing with the volume teardown.

**Comment by torredil** [comment]
@rphillips  the change introduced in this PR does not alter when volume teardown is initiated. It simply implements a best effort mechanism to wait for teardown to complete before the inhibitor lock is released. 

Also, because volume teardown for any given pod only begins when all its containers have stopped, that check is not necessary:

https://github.com/kubernetes/kubernetes/blob/611cbbf64b6ef8b73bb4a1f6dc275bef28287719/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L240

https://github.com/kubernetes/kubernetes/blob/37e73b419e455db34f5fe3e8d815418680ab23df/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L263

(see where reconciler ensures that the pod no longer exists in DSW before issuing an unmount operation):
https://github.com/kubernetes/kubernetes/blob/611cbbf64b6ef8b73bb4a1f6dc275bef28287719/pkg/kubelet/volumemanager/reconciler/reconciler_common.go#L150

**Comment by rphillips** [comment]
Thank you for the clarification.

**Comment by rphillips** [comment]
lgtm
cc @SergeyKanzhelev for an approval

**Comment by torredil** [comment]
For `pull-kubernetes-e2e-gce` all the failures appear to be flakes - see this previous run (same commit) where only the e2e test implemented in this PR fails: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125070/pull-kubernetes-e2e-gce/1811767981395415040:
```
Kubernetes e2e suite: [It] [sig-storage] [NodeFeature:GracefulNodeShutdown] Pod that uses a persistent volume via gce pd driver should get immediately rescheduled to a different node after graceful node shutdown 
```

https://github.com/kubernetes/test-infra/pull/32728 needs to be merged so that a) we can run the e2e test and b) it is skipped for jobs which do not have `GracefulNodeShutdown` enabled.

**Comment by torredil** [comment]
Thanks all for the reviews. I've decided to remove the e2e test from this PR for 2 reasons:

1. It unblocks progress and allows for merging the fix without being blocked by test-infra considerations.
2. It eliminates the need for adding a new job that will only run a single, optional e2e test that may not provide additional signals.

Unit tests remain and I've manually validated the changes. If there's strong interest in e2e testing for this code path, we can consider adding it in a followup PR after the necessary test-infra changes are in place.

**Comment by gnufied** [comment]
@torredil can you summarize, why you think previously failing test was not a issue and we are good to merge this PR now. So `GracefulNodeShutdown` feature is a beta feature which is also enabled by default and I would have thought any test that verifies its behavior should pass by default.

**Comment by torredil** [comment]
> _hmm, it is hard to say now, if the e2e issue was real or indeed a flake._

No changes were made besides removing the e2e test, you can take a look at the diff here: https://github.com/kubernetes/kubernetes/compare/d9e9cefb2fcfbd848b41b391379bd9dde0c923c0..bae995bfb32d7b87f0ed5c7ba2f1da50598bddc5 - `pull-kubernetes-e2e-gce` is green.

> _can you summarize, why you think previously failing test was not a issue and we are good to merge this PR now. So GracefulNodeShutdown feature is a beta feature which is also enabled by default and I would have thought any test that verifies its behavior should pass by default._

The previously failing test was the one that was implemented in this PR. That test requires a new job that configures the needful Kubelet parameters to enable the Graceful Node Shutdown feature. It is correct that `GracefulNodeShutdown` is a beta feature, but to clarify, only the feature gate is enabled by default, not the feature itself. See this note from the [public docs](https://kubernetes.io/blog/2021/04/21/graceful-node-shutdown-beta/#how-do-i-use-it):

- **Note that by default, both configuration options described above, ShutdownGracePeriod and ShutdownGracePeriodCriticalPods are set to zero, so you will need to configure them as appropriate for your environment to activate graceful node shutdown functionality.**

That is to say, there are no jobs which have `GracefulNodeShutdown` enabled today, and thus the e2e test will not succeed because when we send the shutdown signal in the test the nodes will not be gracefully terminated. I opened a PR which does allow us to configure these Kubelet parameters: https://github.com/kubernetes/kubernetes/pull/125413 and that is how we would enable the feature in the new job: https://github.com/kubernetes/test-infra/pull/32728. We would have to run the e2e in that environment for it to succeed.

>  _Are we are not planning to bring back e2es at all?_

We can bring it back if there is enough interest for it, I don't feel strongly about it. As I understand it, folks aren't confident that the test provides a meaningful signal and it requires creating a new job. I believe the current sentiment is that we are okay with creating the job to exercise the optional e2e test, and remove it doesn't end up being all that useful. 

> _My apologize if I am missing something. I will be happy to re-add the lgtm label._

All good these are great questions, let me know if any of the above is not clear or needs more details, there is a quite a bit of context needed to understand this fully and I may have missed some of it.

**Comment by torredil** [comment]
cc @gnufied ^^

**Comment by gnufied** [comment]
>It is correct that GracefulNodeShutdown is a beta feature, but to clarify, only the feature gate is enabled by default, not the feature itself. 

cool. That explains it. 

/lgtm

**Comment by kwilczynski** [comment]
Anything left to do here?

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go [hunk 5]
```diff
@@ -395,19 +400,43 @@ func (m *managerImpl) processShutdownEvent() error {
 			}(pod, group)
 		}
 
+		var groupTerminationWaitDuration = time.Duration(group.ShutdownGracePeriodSeconds) * time.Second
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:403]: can you please add a note here in the code comment to explain why 1 second is added

Comment (SergeyKanzhelev) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:403]: (this is retroactive improvement of code)

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go [hunk 6]
```diff
@@ -395,19 +400,43 @@ func (m *managerImpl) processShutdownEvent() error {
...
+				}
+
+				// Waiting for volume teardown is done on a best basis effort,
+				// report an error and continue.
+				//
+				// Depending on the user provided kubelet configuration value
+				// either the `timer` will tick and we'll continue to shutdown the next group, or,
+				// WaitForAllPodsUnmount will timeout, therefore this goroutine
+				// will close doneCh and we'll continue to shutdown the next group.
+				m.logger.V(1).Error(err, "Failed while waiting for all the volumes belonging to Pods in this group to unmount", "pods", podNames)
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:431]: do you need `V(1)` here? Just `Error` must be fine unless I am missing something

Comment (SergeyKanzhelev) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:431]: is there any test covering the cancellation on a timeout?

File: pkg/kubelet/volumemanager/volume_manager.go [hunk 7]
```diff
@@ -479,6 +487,25 @@ func (vm *volumeManager) WaitForUnmount(ctx context.Context, pod *v1.Pod) error
...
+	var (
+		errors []error
+		wg     sync.WaitGroup
+	)
+	wg.Add(len(pods))
+	for _, pod := range pods {
+		go func(pod *v1.Pod) {
+			defer wg.Done()
+			if err := vm.WaitForUnmount(ctx, pod); err != nil {
+				klog.V(5).ErrorS(err, "Failed to wait for Pod to unmount its volumes", "pod", klog.KObj(pod))
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/volumemanager/volume_manager.go:500]: Why `V(5).Error`? Are you thinking it will be so often that we want to hide this error in a default logging level?

File: pkg/kubelet/volumemanager/volume_manager_test.go [hunk 8]
```diff
@@ -545,3 +549,77 @@ func delayClaimBecomesBound(
...
+				podManager.AddPod(pod)
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+
+			err := manager.WaitForAllPodsUnmount(ctx, test.pods)
+			if test.expectedErrors == nil {
+				if err != nil {
+					t.Errorf("Expected no error, got: %v", err)
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/volumemanager/volume_manager_test.go:611]: nit: there is an assert for this

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go (hunk 5 above)
Comment (torredil) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:405]: Added comment.

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go (hunk 6 above)
Comment (torredil) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go]: `V(1)` is not needed here. Just `Error` is indeed fine, fixed.

Comment (torredil) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go]: Added this code coverage in the latest rev.

File: pkg/kubelet/volumemanager/volume_manager.go (hunk 7 above)
Comment (torredil) [reply pkg/kubelet/volumemanager/volume_manager.go]: Removed this log statement in the newest rev - the error gets returned to the caller and logged there at the default logging level.

File: pkg/kubelet/volumemanager/volume_manager_test.go (hunk 8 above)
Comment (torredil) [reply pkg/kubelet/volumemanager/volume_manager_test.go]: Thanks! latest rev uses assert.

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go (hunk 5 above)
Comment (SergeyKanzhelev) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:405]: I was mostly asking why we are adding 1 second to the configured duration. Can you please add this comment

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go [hunk 9]
```diff
@@ -395,19 +400,45 @@ func (m *managerImpl) processShutdownEvent() error {
...
 			wg.Wait()
+
+			// The signal to kill a Pod was sent successfully to all the pods,
+			// let's wait until all the volumes are unmounted from all the pods before
+			// continuing to the next group. This is done so that the CSI Driver (assuming
+			// that it's part of the highest group) has a chance to perform unmounts.
+			if err := m.volumeManager.WaitForAllPodsUnmount(ctx, group.Pods); err != nil {
+				var podNames []string
+				for _, pod := range group.Pods {
+					podNames = append(podNames, pod.Name)
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:423]: do we need pod namespace in this error message as well?

File: pkg/kubelet/volumemanager/volume_manager.go [hunk 10]
```diff
@@ -479,6 +487,24 @@ func (vm *volumeManager) WaitForUnmount(ctx context.Context, pod *v1.Pod) error
...
+func (vm *volumeManager) WaitForAllPodsUnmount(ctx context.Context, pods []*v1.Pod) error {
+	var (
+		errors []error
+		wg     sync.WaitGroup
+	)
+	wg.Add(len(pods))
+	for _, pod := range pods {
+		go func(pod *v1.Pod) {
+			defer wg.Done()
+			if err := vm.WaitForUnmount(ctx, pod); err != nil {
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/volumemanager/volume_manager.go:499]: if `WaitForUnmount` times out - does it return error or exist with `nil`?

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go [hunk 11]
```diff
@@ -395,19 +400,45 @@ func (m *managerImpl) processShutdownEvent() error {
...
 		go func() {
 			defer close(doneCh)
+			defer ctxCancel()
 			wg.Wait()
+
+			// The signal to kill a Pod was sent successfully to all the pods,
+			// let's wait until all the volumes are unmounted from all the pods before
+			// continuing to the next group. This is done so that the CSI Driver (assuming
+			// that it's part of the highest group) has a chance to perform unmounts.
+			if err := m.volumeManager.WaitForAllPodsUnmount(ctx, group.Pods); err != nil {
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:420]: if timed out - will it exit with error? If not - should we somehow have a log statement about it?

File: pkg/kubelet/volumemanager/volume_manager_test.go [hunk 12]
```diff
@@ -545,3 +550,65 @@ func delayClaimBecomesBound(
...
+				manager)
+
+			err := manager.WaitForAttachAndMount(context.Background(), pod)
+			if err != nil {
+				t.Fatalf("Failed to wait for attach and mount: %v", err)
+			}
+
+			err = manager.WaitForAllPodsUnmount(ctx, []*v1.Pod{pod})
+
+			if test.expectedError {
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/volumemanager/volume_manager_test.go:606]: can we also validate that the duration of execution is not too long when we timed out and matches the timeout we passed?

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go (hunk 6 above)
Comment (SergeyKanzhelev) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go]: Can you please point me to it? I am looking for a test that validates that we are passing all the contexts correctly and the group processing will exit in given time

File: pkg/kubelet/volumemanager/volume_manager.go (hunk 10 above)
Comment (torredil) [reply pkg/kubelet/volumemanager/volume_manager.go:499]: If `WaitForUnmount` times out and there are still volumes mounted, it returns an error. If the function times out but all volumes have been unmounted it returns `nil`:

https://github.com/kubernetes/kubernetes/blob/dd4943c831dbc43fb8d410ccdbc27e3c807131fa/pkg/kubelet/volumemanager/volume_manager.go#L468-L470

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go (hunk 11 above)
Comment (torredil) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:420]: If any of the `WaitForUnmount` calls time out and return an error (which happens when volumes are still mounted after the timeout), those errors will be included in the aggregate error, which is already logged.

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go (hunk 9 above)
Comment (torredil) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go]: Added pod namespace in the error message.

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go (hunk 6 above)
Comment (torredil) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go]: `Test_processShutdownEvent_VolumeUnmountTimeout` in the newest revision.

**Comment by torredil** [comment]
@SergeyKanzhelev This is ready for your eyes again. Please let me know if you spot any further improvements that can be made here and I will be happy to make those changes.

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go (hunk 5 above)
Comment (SergeyKanzhelev) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:405]: sorry, going blind here.

**Review by SergeyKanzhelev — COMMENTED** [review]
lgtm. Some notes:

1. Logic seems to be well time-bounded and test for it is added
2. If there is an unexpected error, the worst case scenario is that more logs will be written
3. Logic is scoped under the Graceful shutdown feature gate and only works on node termination

Since there were many changes I would wait till after weekend for another couple of eyes or until there are maybe another lgtm added. Please ping if I will lose track of it

**Comment by kwilczynski** [comment]
@torredil, looks good! Thank you!

**Comment by torredil** [comment]
cc: @mrunalp

**Review by mrunalp — APPROVED** [review]
